### PR TITLE
add extract_dataset function

### LINKED ISSUE #1469: Add helper function to easily stack `chain` and `draws` dimensions
Stacking chains and draws is often useful when one doesn't care about which chain a draw is coming from. This is currently possible by doing `idata.posterior.stack(sample=("chain", "draw"))`, but very few people seem to know that.

Adding a helper function to easily stack the `chain` and `draws` dimensions into a `sample` dimension would go a long way towards making users' life easier, as well as making better use of `xarray`'s capabilities. Feel free to comment if you want to take on this issue 🖖

### PR DESCRIPTION
**Pull request by OriolAbril**
## Description
Fixes #1469. I think this will cover most casesin the issue and several
more common enough to be relevant like random subsetting for plotting for example.

I probably won't be able to work on this and finish it in the immediate future, if we want this to
be included in the next release someone else should take over, add tests and some examples in the
docstring. If interested comment and go for it, feel free to open a pr to this branch, create a
new PR that builds on top of this work, or work directly on this (if you have permissions to do so).

## Checklist

- [x] Does the PR follow [official](https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md#pull-request-checklist) PR format?
- [x] Is the new feature properly documented with an example? 
- [x] Does the PR include new or updated tests to cover the new feature (using [pytest fixture pattern](https://docs.pytest.org/en/latest/fixture.html#fixture))?
- [x] Is the code style correct (follows pylint and black guidelines)?
- [ ] Is the new feature listed in the [New features](https://github.com/arviz-devs/arviz/blob/main/CHANGELOG.md#new-features)
      section of the changelog?

### CODE REVIEW

**Comment by ahartikainen** [comment]
This function should be able to combine data-arrays from different groups if needed?

**Comment by OriolAbril** [comment]
that could be cool too, how would you indicate that? instead of group and var_names arguments have a group_vars one that is a dict of group names as keys lists of variables names as values?

**Comment by ahartikainen** [comment]
That sounds reasonable option to go with.

**Comment by OriolAbril** [comment]
will review and update this week

File: arviz/utils.py [hunk 1]
```diff
@@ -11,11 +11,84 @@
...
+    var_names = _var_names(var_names, data, filter_vars)
+    if var_names is not None:
+        data = data[var_names]
+    if combined:
+        data = data.stack(__sample__=("chain", "draw"))
+    # 0 is a valid seed se we need to check for rng being exactly boolean
+    if rng is not False:
+        if rng is True:
+            rng = np.random.default_rng()
+        elif isinstance(rng, int):
```
[the remark this answers is not in the document]
Comment (OriolAbril) [reply arviz/utils.py:81]: modified this to a try except, turns out default_rng also takes sequence of integers plus maybe someone uses a subclassed generator. Catched two errors to hopefully provide more informative errors.

File: arviz/utils.py [hunk 2]
```diff
@@ -11,11 +11,84 @@
...
+            rng = np.random.default_rng(rng)
+        if not isinstance(rng, np.random.Generator):
+            raise ValueError("Could not interpret rng as np.random.Generator")
+        random_subset = rng.permutation(np.arange(len(data["__sample__"])))
```
[the remark this answers is not in the document]
Comment (OriolAbril) [reply arviz/utils.py:85]: The results are only sorted if 
* all samples are returned, stacked or not. I assumed in this case, users are keeping all the samples to use all of them, not to cut them right after that. But we can change the default for the stacked case though.
* `rng=False` explicitly (not the default)  

In all other cases, the result is in this `random_subset` order. For example:

```
In [2]: az.extract_dataset(idata, combined=True, var_names="home", num_samples=4)
Out[2]: 
<xarray.Dataset>
Dimensions:  (sample: 4)
Coordinates:
  * sample   (sample) MultiIndex
  - chain    (sample) int64 3 2 1 0
  - draw     (sample) int64 36 390 375 496
Data variables:
    home     (sample) float64 0.1911 0.1604 0.1791 0.1109
Attributes:
    created_at:                 2019-07-12T20:31:53.545143
    inference_library:          pymc3
    inference_library_version:  3.7
```

**Comment by OriolAbril** [comment]
Should be ready to merge now, but still happy to have reviews!

**Review by aloctavodia — APPROVED** [review]
LGTM! Remember to update the changelog

**Pull request merged**